### PR TITLE
Save context manager

### PR DESCRIPTION
**Pull request by typemytype**

### CODE REVIEW

File: drawBot/drawBotDrawingTools.py [hunk 1]
```diff
@@ -465,6 +465,16 @@ def save(self):
...
+        class SaveContextManager(object):
+
+            def __enter__(saveCxt):
+                return saveCxt
+
+            def __exit__(saveCxt, type, value, traceback):
+                self.restore()
+
+        return SaveContextManager()
+
```
Comment (justvanrossum) [inline drawBot/drawBotDrawingTools.py]: I would write it like so:
```python
# at global level
class SaveContextManager(object):
    def __init__(self, drawingTools):
        self._drawingTools = drawingTools
    def __enter__(self):
        return self
    def __exit__(self type, value, traceback):
        self._drawingTools.restore()

# in save():
    return SaveContextManager(self)
```

Comment (typemytype) [reply drawBot/drawBotDrawingTools.py]: I didnt want to keep a reference of a callback or an object somewhere...
this feels very similar as a decorator wrapper

Comment (justvanrossum) [reply drawBot/drawBotDrawingTools.py]: Why not? To not have `self` as a first argument is very ugly. It's an explicit reference as opposed to a scoped variable, which I think is rather obscure when using a nested class.

**Comment by justvanrossum** [comment]
Trying to document this and writing an doc string example makes me doubt the name again. Let's sleep on it, perhaps a different name is easier to explain. Would `savedState()` be an option?

```python
rect(0, 0, 100, 100)
with savedState():
    translate(100, 100)
    rotate(45)
    rect(0, 0, 100, 100)
rect(100, 100, 100, 100)
```

**Comment by justvanrossum** [comment]
A different name makes it easier to refer to good old save/restore and vice versa. Eventually save/restore may be deprecated, but that will be harder if we overload save() with a context manager.

**Comment by justvanrossum** [comment]
I slept over it and am now strongly in favor of `savedState()`. I can implement that later, unless you beat me to it.

File: drawBot/drawBotDrawingTools.py [hunk 2]
```diff
@@ -496,6 +496,26 @@ def restore(self):
...
+        """
+        Save and restore the current state in a `with` statement.
+
+        .. downloadcode:: savedState.py
+
+            # use with statement
+            # this will be wrapped around a `save` and `restore`
+            with savedState():
+                # set a color
+                fill(1, 0, 0)
```
Comment (justvanrossum) [inline drawBot/drawBotDrawingTools.py]: I would love a transformation in the example, too, like rotate and/or translate. I hardly ever use (or need) a saved graphics state for paint styles, but all the time for transformations.

**Comment by justvanrossum** [comment]
```
======================================================================
ERROR: test_rect (testExamples.ExampleTester)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/just/code/git/drawbot/tests/testExamples.py", line 143, in test
    self.assertFilesEqual(imagePath, expectedImagePath)
  File "/Users/just/code/git/drawbot/tests/testExamples.py", line 69, in assertFilesEqual
    self.assertEqual(readData(path1), readData(path2), "Files %r and %s are not the same" % (path1, path2))
  File "/Users/just/code/git/drawbot/tests/testSupport.py", line 119, in readData
    with open(path, "rb") as f:
IOError: [Errno 2] No such file or directory: '/Users/just/code/git/drawbot/tests/data/example_rect.png'

======================================================================
FAIL: test_lineDash (testExamples.ExampleTester)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/just/code/git/drawbot/tests/testExamples.py", line 143, in test
    self.assertFilesEqual(imagePath, expectedImagePath)
  File "/Users/just/code/git/drawbot/tests/testExamples.py", line 69, in assertFilesEqual
    self.assertEqual(readData(path1), readData(path2), "Files %r and %s are not the same" % (path1, path2))
AssertionError: Files '/Users/just/code/git/drawbot/tests/tempTestData/example_lineDash.png' and /Users/just/code/git/drawbot/tests/data/example_lineDash.png are not the same
```

**Comment by typemytype** [comment]
does this one needs more editing?

perhaps test could be added later in an other PR

**Pull request merged**